From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PolishKrowa <45701824+Mrredstone5230@users.noreply.github.com>
Date: Fri, 19 Jul 2024 22:37:06 -0400
Subject: [PATCH] Adding allow-list in moose config

This allows non-vanilla commands specified in moose config to be used in command blocks.

diff --git a/src/main/java/com/mojang/brigadier/tree/CommandNode.java b/src/main/java/com/mojang/brigadier/tree/CommandNode.java
index 89602810e919a92fa7e771c1b443605afaa0c82d..cfd5be45cb93eb351889d2bcd9853a6d80e98373 100644
--- a/src/main/java/com/mojang/brigadier/tree/CommandNode.java
+++ b/src/main/java/com/mojang/brigadier/tree/CommandNode.java
@@ -4,6 +4,7 @@
 package com.mojang.brigadier.tree;
 
 // CHECKSTYLE:OFF
+
 import com.mojang.brigadier.AmbiguityConsumer;
 import com.mojang.brigadier.Command;
 import com.mojang.brigadier.RedirectModifier;
@@ -24,6 +25,12 @@ import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.function.Predicate;
 import net.minecraft.commands.CommandSourceStack;
+// Moose start
+import com.legitimoose.lsp.MooseConfig;
+import net.minecraft.commands.Commands;
+import net.minecraft.server.MinecraftServer;
+import org.bukkit.Bukkit;
+// Moose end
 
 public abstract class CommandNode<S> implements Comparable<CommandNode<S>> {
     private final Map<String, CommandNode<S>> children = new LinkedHashMap<>();
@@ -104,6 +111,13 @@ public abstract class CommandNode<S> implements Comparable<CommandNode<S>> {
             }
         } else {
             this.children.put(node.getName(), node);
+            // Moose start - Allow extra commands from moose config
+            if (MinecraftServer.getServer() != null && MinecraftServer.getServer().getCommands().getDispatcher().getRoot().equals(this) && MooseConfig.allowedExtraCommands.contains(node.getName())) {
+                node.commandBlockAllowed = true;
+                Commands.vanillaCommands.getRoot().addChild((CommandNode<CommandSourceStack>) node);
+                Bukkit.getLogger().info("Allowed extra command " + node.getName());
+            }
+            // Moose end
             if (node instanceof LiteralCommandNode) {
                 this.literals.put(node.getName(), (LiteralCommandNode<S>) node);
             } else if (node instanceof ArgumentCommandNode) {
